### ISSUE TITLE
Improving the front-end codebase

### ISSUE THREAD

**Issue by querkmachine** [Weasyl/weasyl#38]
Hey there,

I'm interested in contributing to Weasyl, but I'm not a Python developer. What I am is a front-end developer and UX designer. 

I figured I'd tackle #19 while I had some free time, but upon looking at the code used for the front-end it all seems a little archaic—at least from my perspective. I didn't want to jump right in and start changing huge swathes of code (especially as I've never contributed before), so I'm opening it to the floor to see if there's much appetite for refactoring.
- Having nearly all of the CSS and JavaScript combined into two files is messy and can be daunting and confusing to look through. Separating these out into partials then automatically concatenating/minifying would make for tidier code and smaller filesizes being sent through to the browser. 
- Relatedly, using a preprocessor for CSS compilation would increase DRY-ness and aid in creating partials. Personally I would recommend [Sass](http://sass-lang.com/) for this, given it's by far the most widespread and feature-rich preprocessor there is.
- Manually managing vendor prefixes (a related oversight of which has caused [at least one bug](https://forums.weasyl.com/vbulletin/showthread.php?8667-KitKat-browser-thumbnail-bug) recently) is tedious and easily automated with tools like [Autoprefixer](https://github.com/postcss/autoprefixer). This additionally makes it easier to use new technologies like Flexbox as it manages the ridiculous number of backwards-compatible vendor prefixes for you.
- A lot of front-end code standards and recommendations aren't really being followed. The amount of IDs being used in selectors especially boggles my mind. FYI using IDs in CSS is discouraged because it throws [selector specificity](https://stuffandnonsense.co.uk/archives/css_specificity_wars.html) [and the cascade](http://csswizardry.com/2014/10/the-specificity-graph/) out of whack and discourages code reuse. 

In my own projects and those I do for my company we'd typically implement this stuff using Node and Gulp, but as I'm unfamiliar with the Weasyl code I'm open to suggestions for how this could be done, if there's much interest in doing it at all. 

Let me know what you think!
~ Kim

**Comment by weykent** [comment]
>Having nearly all of the CSS and JavaScript combined into two files is messy and can be daunting and confusing to look through. Separating these out into partials then automatically concatenating/minifying would make for tidier code and smaller filesizes being sent through to the browser.
>Relatedly, using a preprocessor for CSS compilation would increase DRY-ness and aid in creating partials. Personally I would recommend Sass for this, given it's by far the most widespread and feature-rich preprocessor there is.

This is definitely a thing that should be changed. There was some momentum in fixing this by making an alternate frontend to Weasyl which had a real, full asset pipeline. It never got backported. But, since the effort seems to have been abandoned, backporting makes sense.

>Manually managing vendor prefixes (a related oversight of which has caused at least one bug recently) is tedious and easily automated with tools like Autoprefixer. This additionally makes it easier to use new technologies like Flexbox as it manages the ridiculous number of backwards-compatible vendor prefixes for you.

Absolutely. There's a lot of changes that should be made.

>A lot of front-end code standards and recommendations aren't really being followed. The amount of IDs being used in selectors especially boggles my mind. FYI using IDs in CSS is discouraged because it throws selector specificity and the cascade out of whack and discourages code reuse.

It is in dire need of refactoring/rewriting if it's not going to be thrown out, which was the original plan. If you want to help with this project, it would be greatly appreciated. I don't believe the original frontend author has been around much lately, but we can see if this works: hey, @adenzerda!!

>In my own projects and those I do for my company we'd typically implement this stuff using Node and Gulp, but as I'm unfamiliar with the Weasyl code I'm open to suggestions for how this could be done, if there's much interest in doing it at all.

Absolutely. The aforementioned bits were done with grunt (several years ago), but I guess gulp is largely replacing it these days.

I'll post the remaining code so it can be referenced. There's a number of other things that would be useful to backport anyway.

**Comment by weykent** [comment]
https://github.com/Weasyl/weasyl3 Check it.

**Comment by querkmachine** [comment]
Looking better! Happy to see that many of my raised issues were already being tackled, even if the attempt was ultimately abandoned. 

I'm pretty interested in helping out, though my unfamiliarity with the code and ever-present day job may make for slow progress. Still, I'll pull down the repo and fiddle around with it (and hopefully not break anything!), something magical might come out of it!

**Comment by charmander** [comment]
I agree, but would just like to make a note about this common advice:

> FYI using IDs in CSS is discouraged because it throws selector specificity and the cascade out of whack and discourages code reuse.

ids aren’t bad, just often misused. It’s not as if you can’t put both an id and several classes on an element, and any rules involving them would ideally be placed the same in the cascade regardless of specificity.

**Comment by weykent** [comment]
@querkmachine If you don't mind, could you hop on [our gitter chat](https://gitter.im/Weasyl/weasyl) to talk about possible approaches? Just to make sure we're all on the same page and so on.

**Comment by charmander** [comment]
#46 is in now, which should make it possible to continue cleaning things up without deployment changes.

**Comment by querkmachine** [comment]
Well daymn, that's looking nice. 

Unfortunately between work and jetlag I can't claim to have made much progress with actually untangling the big ol' CSS file. From what I can tell there's no easy way of splitting it up into sensible partials without also changing the classes/IDs in the templates—which then has a potential knock-on effect in the JavaScript—or starting to duplicate/mixin-ize significant chunks of code. 

It's technically possible to do it without touching the HTML/JS but it's going to come out messier than it needs to be if you don't, IMO.

**Comment by charmander** [comment]
I don’t think we ever wanted to do it without changing the templates; they’re in need of just as much cleanup.

**Comment by querkmachine** [comment]
Is there a means of having HTML changes show up without having to kill and re-run `make guest-run`?It's pretty frustrating having to run a command every time something has changed. 

This also extends to the file revisioning introduced in #46, as the compiled CSS file's name is changed with each modification, however the HTML is stuck requesting the file that existed when that iteration of `make guest-run` was first run, which buggers things up, as you might imagine.

**Comment by charmander** [comment]
#82 should help with that! I’ve been using it locally but never submitted a pull request, whoops.
